src/gui_textual/connection_services.py

```diff
@@ -0,0 +1,60 @@
+"""The business logic of connecting and communicating with devices."""
+
+from zaber_motion.ascii import Connection, Device, Axis
+from zaber_motion.exceptions import ConnectionFailedException, CommandFailedException
+
+
+class ConnectionServices:
+    """Provide connection services to devices in a singleton."""
+
+    def __init__(self) -> None:
+        """Initialize ConnectionServices."""
+        self._connection: Connection | None = None
+        self._device_list: list[Device] = []
+        self._device: Device | None = None
+        self._axis: Axis | None = None
+
+    def open_serial(self, port: str) -> str:
+        """Open a serial port."""
+        self.close()
+        try:
+            self._connection = Connection.open_serial_port(port)
+        except ConnectionFailedException:
+            return "Unable to open serial port."
+        self._connection.enable_alerts()
+        self._device_list = self._connection.detect_devices()
+        self._device = self._device_list[0]
+        self._axis = self._device.get_axis(1)
+        return f"Connected to {str(self._connection)}"
```

Comment:
I think this `str()` call is redundant

```suggestion
        return f"Connected to {self._connection}"
```